lib/logstash/outputs/webhdfs.rb

```diff
@@ -0,0 +1,262 @@
+    newline = "\n"
+    output_files = Hash.new { |hash, key| hash[key] = "" }
+    events.collect do |event|
+      path = event.sprintf(@path)
+      if @message_format
+        event_as_string = event.sprintf(@message_format)
+      else
+        event_as_string = event.to_json
+      end
+      event_as_string += newline unless event_as_string.end_with? newline
+      output_files[path] << event_as_string
+    end
+    output_files.each do |path, output|
+      if @compression == "gzip"
+        path += ".gz"
+        output = compress_gzip(output)
+      elsif @compression == "snappy"
+        path += ".snappy"
+        if @snappy_format == "file"
+          output = compress_snappy_file(output)
+        elsif
+          output = compress_snappy_stream(output)
+        end
+      end
+      write_data(path, output)
+    end
+  end
+
+  def compress_gzip(data)
+    buffer = StringIO.new('','w')
+    compressor = Zlib::GzipWriter.new(buffer)
+    begin
+      compressor.write(data)
+    ensure
+      compressor.close()
+    end
+    buffer.string
+  end
+
+  def compress_snappy_file(data)
+    # Encode data to ASCII_8BIT (binary)
+    data= data.encode(Encoding::ASCII_8BIT, "binary", :undef => :replace)
+    buffer = StringIO.new('', 'w')
+    buffer.set_encoding(Encoding::ASCII_8BIT)
+    compressed = Snappy.deflate(data)
+    buffer << [compressed.size, compressed].pack("Na*")
+    buffer.string
+  end
+
+  def compress_snappy_stream(data)
+    # Encode data to ASCII_8BIT (binary)
+    data= data.encode(Encoding::ASCII_8BIT, "binary", :undef => :replace)
+    buffer = StringIO.new
+    buffer.set_encoding(Encoding::ASCII_8BIT)
+    chunks = data.scan(/.{1,#{@snappy_bufsize}}/m)
+    chunks.each do |chunk|
+      compressed = Snappy.deflate(chunk)
+      buffer << [chunk.size, compressed.size, compressed].pack("NNa*")
+    end
+    return buffer.string
+  end
+
+  def get_snappy_header!
+    [MAGIC, DEFAULT_VERSION, MINIMUM_COMPATIBLE_VERSION].pack("a8NN")
+  end
+
+  def write_data(path, data)
+    # Retry max_retry times. This can solve problems like leases being hold by another process. Sadly this is no
+    # KNOWN_ERROR in rubys webhdfs client.
+    write_tries = 0
+    begin
+      # Try to append to already existing file, which will work most of the times.
+      @client.append(path, data)
+    # File does not exist, so create it.
+    rescue WebHDFS::FileNotFoundError
+      # Add snappy header if format is "file".
+      if @compression == "snappy" and @snappy_format == "file"
+        @client.create(path, get_snappy_header! + data)
+      elsif
+        @client.create(path, data)
+      end
+    # Handle other write errors and retry to write max. @retry_times.
+    rescue => e
+      if write_tries < @retry_times
+        @logger.warn("webhdfs write caused an exception: #{e.message}. Maybe you should increase retry_interval or reduce number of workers. Retrying...")
+        sleep(@retry_interval * write_tries)
```

Comment:
If event throughput is high and the hdfs cluster is not able to handle that load, this problem might get triggered. It is caused by to many concurrent writes to the same file. The retry here is only mend to solve problems with short spikes in event throughput. I did not want to introduce yet another dependency for this plugin, thus the rather naive backoff mechanism. By our experience it did work, though.